include/swift/module.modulemap

```diff
@@ -0,0 +1,157 @@
+// This is defined as one big module because there are cyclic dependencies
+// between the submodules all over the place. Breaking this up may actually
+// speed up compilation.
+
+module Swift {
+  // FIXME: Define submodules for all the other directories.
+
+  module Strings {
+    requires cplusplus
+    header "Strings.h"
+    export *
+  }
+
+  module ABI {
+    requires cplusplus
+    // No submodules, there are missing includes.
+    umbrella "ABI"
+   
+    // These headers introduce a cyclic dependency with Swift_Shims.
+    textual header "ABI/HeapObject.h"
+    textual header "ABI/KeyPath.h"
+    textual header "ABI/Metadata.h"
+    textual header "ABI/MetadataValues.h"
+    textual header "ABI/System.h"
+   
+    textual header "ABI/MetadataKind.def"
+    textual header "ABI/ValueWitness.def"
+  }
+
+  module AST {
+    requires cplusplus
+    umbrella "AST" //module * { export * }
+    textual header "AST/ReferenceStorage.def"
+    textual header "AST/StmtNodes.def"
+    textual header "AST/AccessTypeIDZone.def"
+    textual header "AST/TypeCheckerTypeIDZone.def"
+    textual header "AST/AccessorKinds.def"
+    textual header "AST/TypeNodes.def"
+    textual header "AST/Attr.def"
+    textual header "AST/TypeReprNodes.def"
+    textual header "AST/Builtins.def"
+    textual header "AST/DeclNodes.def"
+    textual header "AST/DiagnosticsAll.def"
+    textual header "AST/DiagnosticsClangImporter.def"
+    textual header "AST/DiagnosticsCommon.def"
+    textual header "AST/DiagnosticsDriver.def"
+    textual header "AST/DiagnosticsFrontend.def"
+    textual header "AST/DiagnosticsIRGen.def"
+    textual header "AST/DiagnosticsModuleDiffer.def"
+    textual header "AST/DiagnosticsParse.def"
+    textual header "AST/DiagnosticsRefactoring.def"
+    textual header "AST/DiagnosticsSIL.def"
+    textual header "AST/DiagnosticsSema.def"
+    textual header "AST/ExprNodes.def"
+    textual header "AST/KnownDecls.def"
+    textual header "AST/KnownFoundationEntities.def"
+    textual header "AST/KnownIdentifiers.def"
+    textual header "AST/KnownProtocols.def"
+    textual header "AST/KnownStdlibTypes.def"
+    textual header "AST/NameLookupTypeIDZone.def"
+    textual header "AST/PatternNodes.def"
+    textual header "AST/PlatformKinds.def"
+
+    // Missing include.
+    textual header "AST/DiagnosticsCommon.h"
+    textual header "AST/FileSystem.h"
+
+    // Included inside a namespace.
+    textual header "Basic/DefineTypeIDZone.h"
+    textual header "Basic/ImplementTypeIDZone.h"
```

Comment:
@DougGregor, should these be `.inc` files, or some other extension that makes them not headery?

Reply:
Yes, we can give them a less headery extensions. `.def`, maybe?